packages/lesswrong/lib/notificationTypes.tsx

```diff
@@ -66,23 +76,77 @@ const registerNotificationType = ({allowedChannels = ["none", "onsite", "email",
   return notificationTypeClass;
 }
 
-const getDocument = async (documentType: string|null, documentId: string|null) => {
-  if (!documentId) return null;
-  
-  switch(documentType) {
-    case "post":
-      return await Posts.findOne(documentId);
-    case "comment":
-      return await Comments.findOne(documentId);
-    case "user":
-      return await Users.findOne(documentId);
-    case "message":
-      return await Messages.findOne(documentId);
-    case "tagRel":
-      return await TagRels.findOne(documentId);
+export const getDocument = async (documentType: NotificationDocument | null, documentId: string | null) =>
+  (await getDocumentSummary(documentType, documentId))?.document
+
+type DocumentSummary =
+  | { type: 'post'; associatedUserName: string; displayName: string; document: DbPost }
+  | { type: 'comment'; associatedUserName: string; displayName: string | undefined; document: DbComment }
+  | { type: 'user'; associatedUserName: string; displayName: string; document: DbUser }
+  | { type: 'message'; associatedUserName: string; displayName: string | undefined; document: DbMessage }
+  | { type: 'localgroup'; displayName: string; document: DbLocalgroup; associatedUserName: null }
+  | { type: 'tagRel'; document: DbTagRel; associatedUserName: null; displayName: null }
+
+export const getDocumentSummary = async (documentType: NotificationDocument | null, documentId: string | null): Promise<DocumentSummary | null> => {
```

Comment:
todo: add more on why

Reply:
@jimrandomh seemed probably good for you to take a look at this refactor and see if it made sense to you.

Reply:
Motivation for this was that:
- I needed to handle the case where I have different document types for single notification type, so I need a function like this (though not necessarily covering all the cases)
- Most other notification types follow the pattern of get the title of the thing and the author of the thing and can be rewritten using this function. 
  - Which abstracts away the details of the correct way of getting title/author (rn each notification has to know about and use the proper utility functions for that e.g. using `userGetDisplayName` vs using `.displayName` as a few of the types do rn)
  - Actually probably most of "getMessage" functions can be auto-generated from a template string (with author and title being template parameters)

Plausibly getLink stuff should also be incorporated here to minimize number of times we do switch over this set of values, but that was a different code flow that I preferred not to touch for now.